nexus/src/db/datastore.rs

```diff
@@ -1143,6 +1196,114 @@ impl DataStore {
+    /// and an [`authz::Disk`] for doing authz checks
+    ///
+    /// See [`DataStore::organization_lookup_noauthz()`] for intended use cases
+    /// and caveats.
+    // TODO-security See the note on organization_lookup_noauthz().
+    async fn disk_lookup_noauthz(
+        &self,
+        authz_project: &authz::Project,
+        disk_name: &Name,
+    ) -> LookupResult<(authz::Disk, Disk)> {
+        use db::schema::disk::dsl;
+        dsl::disk
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::project_id.eq(authz_project.id()))
+            .filter(dsl::name.eq(disk_name.clone()))
+            .select(Disk::as_select())
+            .first_async(self.pool())
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
+                        ResourceType::Disk,
+                        LookupType::ByName(disk_name.as_str().to_owned()),
+                    ),
+                )
+            })
+            .map(|d| {
+                (
+                    authz_project.child_generic(
+                        ResourceType::Disk,
+                        d.id(),
+                        LookupType::from(&disk_name.0),
+                    ),
+                    d,
+                )
+            })
+    }
+
+    /// Fetch an [`authz::Disk`] based on its id
+    pub async fn disk_lookup_by_id(
+        &self,
+        disk_id: Uuid,
+    ) -> LookupResult<authz::Disk> {
+        use db::schema::disk::dsl;
+        let project_id = dsl::disk
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(disk_id))
+            .select(dsl::project_id)
+            .first_async(self.pool())
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
+                        ResourceType::Disk,
+                        LookupType::ById(disk_id),
+                    ),
+                )
+            })?;
+        let authz_project = self.project_lookup_by_id(project_id).await?;
+        Ok(authz_project.child_generic(
+            ResourceType::Disk,
+            disk_id,
+            LookupType::ById(disk_id),
+        ))
+    }
+
+    /// Look up the id for a Disk based on its name
+    ///
+    /// Returns an [`authz::Disk`] (which makes the id available).
+    ///
+    /// Like the other "lookup_by_path()" functions, this function does no authz
+    /// checks.
+    // TODO-security For Containers in the hierarchy (like Organizations and
+    // Projects), we don't do an authz check in the "lookup_by_path" functions
+    // because we don't know if the caller has access to do the lookup.  For
+    // leaf resources (like Instances and Disks), though, we do.  We could do
+    // the authz check here, and in disk_lookup_by_id() too.  Should we?
```

Comment:
I worry that having a separation between what's auth'd and what's not auth'd being differentiated by logical hierarchy might potentially lead to bugs. Is there some way we could be more explicit in describing that a particular reference has been auth'd? I mean, we could continue with naming schemes like `_noauthz` or `_no_auth` which is helpful at the callsite... I wonder more if there's something we can do with the return type of these lookups though. One could mistakenly assume that `authz::Disk` meant authorized disk at first glance. Would it be possible to have a type, be it a newtype pattern or otherwise, that wrapped the `authz::*` structs to indicate that they had indeed been authorized. The `authorize` function could be the only thing to return that type and then we could lean on some level of type safety to save us from ourselves.

Reply:
I think it's very likely we'd want more more than an empty new type. My intuition is that we'd want it to be the authz resource reference as well as the actions as you mention. As for what it would look like, I think it'd take some more detailed exploration. If there's a small set of actions with increasing permissions then something using a [bit flag](https://crates.io/crates/flagset) and some logic like anything above a certain value can take an action. I don't _exactly_ know, but having a single field that can represent granted permissions (be it bit flags or otherwise) seems like it'd be something we'd want to include. 

I think my biggest preference for how this would work is that we have a macro that's used at the datastore level that dictates what permissions it takes. That macro would take care of all the auth/validation stuff to ensure the action could be performed. We'd then pair that with a linter to fail anything without explicit macro usage. Something like that anyway. 

I'm not 100% sure, but I definitely think that it's worth exploring. I wouldn't stop the presses to get to that now, it seems more like a pre-launch/post-demo priority.